src/Skylight.Server/Game/Catalog/CatalogOffer.cs

```diff
@@ -44,20 +47,35 @@ internal CatalogOffer(int id, string name, int orderNum, int clubRank, int costC
 		this.HasOffer = hasOffer;
 
 		this.Products = products;
+
+		this.priceCurrency = priceCurrency;
 	}
 
-	public async ValueTask PurchaseAsync(ICatalogTransaction transaction, CancellationToken cancellationToken)
+	private bool CanPurchase(ICatalogTransaction tx) =>
+		tx.GetCurrencyBalance(this.priceCurrency) >= this.CostCredits;
+
+	public async ValueTask PurchaseAsync(
+		ICatalogTransaction transaction,
+		CancellationToken cancellationToken)
 	{
-		foreach (ICatalogProduct product in this.Products)
+		if (this.CostCredits < 0)
```

Comment:
Validating the objects own state like this should be done much earlier. It should be error state on boot so action is taken for invalid configuration and not left as broken that explodes down the road.